docs(delivery): added first draft of e2e ack guide

barieom opened the pull request:
<!--
**Your PR title must conform to the conventional commit spec!**

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

spencergilbert commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,118 @@
+description: Learn how to use end-to-end acknowledgement
+author_github: https://github.com/barieom
+domain: delivery
+weight: 5
```
<img width="787" alt="image" src="https://user-images.githubusercontent.com/11903873/157560843-cc58581d-38a7-4511-937b-f8147a29accb.png">

This might need to be changed, or the others. I assume that's driving the guide numbering

spencergilbert commented on website/content/en/guides/level-up/e2e-ack-guide.md:
````diff
@@ -0,0 +1,118 @@
+data). You can set acknowledgements at the source and sink level by:
+
+```toml
+[sources.cool_source_id]
````
`cool_id` 😆

spencergilbert commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,118 @@
+## Getting Started
+
+You can set and control the acknowledgement feature either at the global level or
+the source level. Let's start with the global configuration option first, as it is
```
I thought we swapped this to be enabled at the `sink` side

jszwedko replied:
Yeah, that is correct, acknowledgements are configured on sinks now instead of sources. When acknowledgements are enabled for a sink, all associated sources that support acknowledgements will wait for delivery. The ones that don't support it, will output a warning message (we should call that out so users know what to look for).

spencergilbert commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,118 @@
+[sources.cool_source_id]
+   # Enable or disable waiting for acknowledgements for this sink.
+   # Defaults to the global value of `acknowledgements`
+   acknowledgement = true
```
Same as above, I thought this was moved to just the `sink`s

spencergilbert commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,118 @@
+Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
+acknowledgement feature. First, as alluded to earlier, not all sources and sinks
+are supported because some sources and sinks are unable to provide acknowledgements
+at the protocol level. A list of sources and sinks that are not supported are:
```
Why did we end up listing components that _don't_ support e2e, rather than listing components that do support it. I'm not sure I feel strongly either way, but I do feel _something_ about it.

jszwedko commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,135 @@
+Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
+acknowledgement feature. First, as alluded to earlier, not all sources and sinks
+are supported because some sources and sinks are unable to provide acknowledgements
+at the protocol level. A list of sources and sinks that _are_ supported are:
```
This list is likely to be out of date quickly. I'd just mention that you can see which components support acknowledgements on each of their component documentation pages by looking for the `acknowledgments` badge.

jszwedko commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,135 @@
+## Parting thoughts
+
+This guide intended to serve as a basic guide on how you can start
+leveraging `acknowledgement`. If you have any feedback regarding this feature,
```
```suggestion
leveraging Vector's end-to-end acknowledgement feature. If you have any feedback regarding this feature,
```

jszwedko reviewed (commented):
Thanks @barieom !

I think this needs to be revised since the configuration was moved from sources to sinks, but otherwise seems good.

barieom commented:
> Thanks @barieom !
> 
> I think this needs to be revised since the configuration was moved from sources to sinks, but otherwise seems good.

Thanks for reviewing @jszwedko ! Addressed all your comments :)

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
````diff
@@ -0,0 +1,86 @@
+configuration below turns end-to-end acknowledgement on for every source:
+
+```toml
+acknowledgement = true
````
Note, `acknowledgement` should be `acknowledgements` everywhere, and the configuration has an `enabled` field. Setting just `acknowledgements = true` works through a compatibility hack.
```suggestion
acknowledgements.enabled = true
```

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,86 @@
+acknowledgement from all the sinks. A source that does acknowledgements will wait
+forever for an ack before responding.
+
+But you can enable acknowledgements at each sink level if you want more granular
```
```suggestion
But you can enable acknowledgements individually for each sink if you want more granular
```

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,86 @@
+
+But you can enable acknowledgements at each sink level if you want more granular
+control over how acknowledgements for specific cases (e.g. you're buffering your
+data). When you enable `acknowledgement` for a supported sink, all associated sources
```
```suggestion
data). When you enable `acknowledgements` for a supported sink, all sources connected
```

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,86 @@
+[sinks.very_cool_id]
+   # Enable or disable waiting for acknowledgements for this sink.
+   # Defaults to the global value of `acknowledgements`
+   acknowledgement = true
```
As above:
```suggestion
   acknowledgements.enabled = true
```

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
````diff
@@ -0,0 +1,86 @@
+```
+
+Even if you have a relatively complex topology and sending data from one source to
+multiple sinks, by enabling this global config, all supported sources will wait for
+acknowledgement from all the sinks. A source that does acknowledgements will wait
+forever for an ack before responding.
````
I'm not sure the latter phrase adds anything to this section. This suggestion is a little awkwardly worded, but comes closer to describing the process.
```suggestion
multiple sinks, by enabling this global config, all sources with acknowledgement support will wait before responding for
acknowledgement from all the sinks that its events flow through.
```

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,86 @@
+Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
+acknowledgement feature. First, as alluded to earlier, not all sources and sinks
+are supported because some sources and sinks are unable to provide acknowledgements
+at the protocol level. Sinks and sources that don't support acknowledgements will
+output a warning message to let you know if you try to enable acknowledgements. You
```
Technically, enabling acknowledgements on a sink that don't support acknowledgements will prevent the configuration from loading (ie fatal error). I don't know if you want to make that distinction.

barieom replied:
addressed

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,86 @@
+## Edge cases
+
+Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
+acknowledgement feature. First, as alluded to earlier, not all sources and sinks
+are supported because some sources and sinks are unable to provide acknowledgements
+at the protocol level. Sinks and sources that don't support acknowledgements will
```
Also, because we haven't implemented support for some yet that do have protocol support.

barieom replied:
addressed

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,86 @@
+pages by looking for the acknowledgments badge.
+
+Second, when buffering your observability data, the behavior of end-to-end
+acknowledgement is a little different. When an event is persisted in a buffer, the
```
Is it worth distinguishing between disk and memory buffers? This behavior is only for disk buffers and I'd hate to cause confusion here.

barieom replied:
addressed

bruceg reviewed (changes requested):
This is a good starting guide. I left some clarifications and comments below.

Do we want to say anything about error feedback (ie what happens if a sink rejects events)?

barieom commented:
> This is a good starting guide. I left some clarifications and comments below.
> 
> Do we want to say anything about error feedback (ie what happens if a sink rejects events)?

Agree that's important. That's now in there.

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,97 @@
+will produce some 400 error code, while sources like Kafka that don't handle
+rejection behaves similarly to a positive acknowledgement.
+
+But you can enable acknowledgements individually for each sink if you want more granular
```
Should the order of these two paragraphs be reversed?

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,97 @@
+
+If any sink rejects a message, that failure marks the batch that the message is
+part of as _failed_, and the source will respond accordingly. For example, HTTP sources
+will produce some 400 error code, while sources like Kafka that don't handle
+rejection behaves similarly to a positive acknowledgement.
```
The phrase "don't handle rejection" sounds awkward.
```suggestion
will produce some 400 error code, while sources like Kafka that don't have protocol support for
rejecting messages behaves similarly to a positive acknowledgement.
```

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,97 @@
+acknowledgement feature.
+
+First, as alluded to earlier, not all sources and sinks are supported. We are, of
+course, working to add support for more sources and sinks, but support for some
+sources and sinks will always be unable because they cannot provide acknowledgements
```
"support … will always be unable" is awkward.

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,97 @@
+configuration from loading. You can find which components support acknowledgements
+on each of their documentation pages by looking for the acknowledgments badge.
+
+Second, when buffering your observability data in a disk, the behavior of end-to-end
+acknowledgement is a little different. When an event is persisted in a disk buffer,
+the event will be marked as acknowledged to the source.
```
Is it worth giving a reason why we do this? ie these events will continue be retried in the event of a crash, so we have accepted responsibility, etc.

bruceg commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,97 @@
+Third, the final edge case is if you are using any user-space transforms, such
+as Lua. Because user-space transforms can drop, combine, split, or create events
+that may have no relation to the source event, tracking whether an event has been
+successfully delivered gets extremely complex. As such, end-to-end acknowledgement
+behavior will be put in the hands of script writers.
```
Actually, unless I'm completely misremembering what I did, the lua transform works much like the disk buffers here, and no acknowledgements are possible.

jszwedko commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,97 @@
+You can set and control the acknowledgement feature either at the global level or
+the sink level. Let's start with the global configuration option first, as it is
+easy as flipping on a switch. As you would expect from a global configuration, the
+configuration below turns end-to-end acknowledgement on for every source:
```
```suggestion
configuration below turns end-to-end acknowledgement on for every sink:
```

bruceg replied:
Well, technically it is turning it on for every source, which is where the behavior change actually happens. This is a point of confusion.

jszwedko commented on website/content/en/guides/level-up/e2e-ack-guide.md:
````diff
@@ -0,0 +1,97 @@
+```
+
+To note, if any sink rejects a message, that failure marks the batch that the message is
+part of as _failed_, and the source will respond accordingly. For example, HTTP sources
+will produce some 400 error code, while sources like Kafka that don't have protocol
+support for rejecting messages behaves similarly to a positive acknowledgement.
````
```suggestion
part of as _failed_, and the source will respond accordingly. For example, HTTP sources
will produce a 400 or 500-level error code, while sources like Kafka that don't have protocol
support for rejecting messages will simply not ack the messages, allowing them to be reprocessed.
```

jszwedko commented on website/content/en/guides/level-up/e2e-ack-guide.md:
```diff
@@ -0,0 +1,97 @@
+we've yet to add support for all sources and sinks that support acknowledgements
+at the protocol level so we'll be adding that overtime.
+
+Sources that don't support acknowledgements will output a
+warning message to let you know if you try to enable acknowledgements, while sinks
```
```suggestion
Sources that don't support acknowledgements will output a
warning message to let you know if you try to enable acknowledgements on a connected sink, while sinks
```

jszwedko reviewed (commented):
Nice work @barieom ! I left a few last comments, but overall this looks good.
